lib/services/local/usertoken.go

```diff
@@ -0,0 +1,182 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package local
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+
+	"github.com/gravitational/trace"
+)
+
+// GetUserTokens returns all user tokens.
+func (s *IdentityService) GetUserTokens(ctx context.Context) ([]types.UserToken, error) {
+	startKey := backend.Key(userTokenPrefix)
+	result, err := s.GetRange(ctx, startKey, backend.RangeEnd(startKey), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// DELETE IN 9.0.0 retrieve tokens with old prefix.
+	startKey = backend.Key(LegacyPasswordTokensPrefix)
+	oldPrefixResult, err := s.GetRange(ctx, startKey, backend.RangeEnd(startKey), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	items := append(result.Items, oldPrefixResult.Items...)
+
+	var tokens []types.UserToken
+	for _, item := range items {
+		if !bytes.HasSuffix(item.Key, []byte(paramsPrefix)) {
+			continue
+		}
+
+		token, err := services.UnmarshalUserToken(item.Value)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		tokens = append(tokens, token)
+	}
+
+	return tokens, nil
+}
+
+// DeleteUserToken deletes user token by ID.
+func (s *IdentityService) DeleteUserToken(ctx context.Context, tokenID string) error {
+	_, err := s.GetUserToken(ctx, tokenID)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	startKey := backend.Key(userTokenPrefix, tokenID)
+	if err = s.DeleteRange(ctx, startKey, backend.RangeEnd(startKey)); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// DELETE IN 9.0.0 also delete any tokens with old prefix.
+	startKey = backend.Key(LegacyPasswordTokensPrefix, tokenID)
+	return trace.Wrap(s.DeleteRange(ctx, startKey, backend.RangeEnd(startKey)))
+}
+
+// GetUserToken returns a token by its ID.
+func (s *IdentityService) GetUserToken(ctx context.Context, tokenID string) (types.UserToken, error) {
+	item, err := s.Get(ctx, backend.Key(userTokenPrefix, tokenID, paramsPrefix))
+	if err != nil {
+		if trace.IsNotFound(err) {
+			// DELETE IN 9.0.0: fallback for old prefix
+			item, err = s.Get(ctx, backend.Key(LegacyPasswordTokensPrefix, tokenID, paramsPrefix))
+			if err != nil {
+				if trace.IsNotFound(err) {
+					return nil, trace.NotFound("user token(%v) not found", tokenID)
+				}
+				return nil, trace.Wrap(err)
+			}
+		} else {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	token, err := services.UnmarshalUserToken(item.Value)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
Suggestion: cut nesting / simplify error handling.  Ditto for all similar blocks in this file.

```suggestion
	item, err := s.Get(ctx, backend.Key(userTokenPrefix, tokenID, paramsPrefix))
	// DELETE IN 9.0.0: fallback for old prefix first.
	if trace.IsNotFound(err) {
		item, err = s.Get(ctx, backend.Key(LegacyPasswordTokensPrefix, tokenID, paramsPrefix))
	}
	// Handle errors from either Get.
	switch {
	case trace.IsNotFound(err):
		return nil, trace.NotFound("user token(%v) not found", tokenID)
	case err != nil:
		return nil, trace.Wrap(err)
	}
```

Reply:
oh this is much better, applied!